File: campus-exchange/server/src/models/Message.js
const mongoose = require('mongoose');

const MessageSchema = new mongoose.Schema(
  {
    messageId: {
      type: String,
      required: true,
      unique: true,
      index: true,
    },
    conversationId: {
      type: String,
      required: true,
      index: true,
      ref: 'Conversation',
    },
    listingId: {
      type: String,
      required: true,
      index: true,
      ref: 'Listing',
    },
    collegeId: {
      type: String,
      required: true,
      index: true,
      ref: 'College',
    },
    senderId: {
      type: String,
      required: true,
      index: true,
      ref: 'Student',
    },
    message: {
      type: String,
      required: [true, 'Message text is required'],
      trim: true,
      maxlength: 1000,
    },
  },
  {
    timestamps: true,
  }
);

MessageSchema.index({ conversationId: 1, createdAt: 1 });
MessageSchema.index({ listingId: 1, createdAt: -1 });

module.exports = mongoose.model('Message', MessageSchema);

File: mobile/lib/models/message.dart
class Message {
  final String messageId;
  final String conversationId;
  final String listingId;
  final String collegeId;
  final String senderId;
  final String message;
  final DateTime createdAt;

  Message({
    required this.messageId,
    required this.conversationId,
    required this.listingId,
    required this.collegeId,
    required this.senderId,
    required this.message,
    required this.createdAt,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      messageId: json['messageId'] ?? '',
      conversationId: json['conversationId'] ?? '',
      listingId: json['listingId'] ?? '',
      collegeId: json['collegeId'] ?? '',
      senderId: json['senderId'] ?? '',
      message: json['message'] ?? '',
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt']) ?? DateTime.now()
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'messageId': messageId,
      'conversationId': conversationId,
      'listingId': listingId,
      'collegeId': collegeId,
      'senderId': senderId,
      'message': message,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}
